statistics/dml_update.go

```diff
@@ -0,0 +1,272 @@
+		}
+		if delta.IdxDeltas[idxInfo.ID] == nil {
+			delta.IdxDeltas[idxInfo.ID] = &HistDelta{}
+		}
+		idxDelta := delta.IdxDeltas[idxInfo.ID]
+		idxDelta.updateHistDelta(sc, &idx.Histogram, value, deltaCount)
+	}
+}
+
+func encodeIndexValue(sc *stmtctx.StatementContext, idxInfo *model.IndexInfo, row []types.Datum) *types.Datum {
+	values := make([]types.Datum, 0, len(idxInfo.Columns))
+	for _, col := range idxInfo.Columns {
+		if col.Offset < 0 || col.Offset >= len(row) {
+			return nil
+		}
+		values = append(values, row[col.Offset])
+	}
+	b, err := codec.EncodeKey(sc, nil, values...)
+	if err != nil {
+		terror.Log(err)
+		return nil
+	}
+	d := types.NewBytesDatum(b)
+	return &d
+}
+
+func (delta *BktDelta) updateBktBound(sc *stmtctx.StatementContext, value *types.Datum, isUpper bool) {
+	if delta.Bound == nil {
+		delta.Bound = value
+		return
+	}
+	cmp, err := delta.Bound.CompareDatum(sc, value)
+	if err != nil {
+		terror.Log(err)
+		return
+	}
+	if (isUpper && cmp == -1) || (!isUpper && cmp == 1) {
+		delta.Bound = value
+	}
+}
+
+func (delta *HistDelta) updateHistDelta(sc *stmtctx.StatementContext, hg *Histogram, value *types.Datum, deltaCount int64) {
+	if hg.Len() == 0 {
+		return
+	}
+	// The histogram version has changed, so we need to migrate the old delta info to the new histogram.
+	if hg.LastUpdateVersion != delta.HistVersion {
+		delta.HistVersion = hg.LastUpdateVersion
+		bktDelta := delta.BktDeltas
+		delta.BktDeltas = nil
+		for _, bkt := range bktDelta {
+			delta.updateHistDelta(sc, hg, bkt.Bound, 0)
+		}
+	}
+	if value == nil {
+		return
+	}
+	idx, match := hg.Bounds.LowerBound(0, value)
+	if delta.BktDeltas == nil {
+		delta.BktDeltas = make(map[int]*BktDelta)
+	}
+	if delta.BktDeltas[idx/2] == nil {
+		delta.BktDeltas[idx/2] = &BktDelta{}
+	}
+	bkt := delta.BktDeltas[idx/2]
+	bkt.Count += deltaCount
+	// This value falls outside the whole histogram, so we need to update the last bucket's upper bound.
+	if idx == hg.Bounds.NumRows() {
+		bkt.updateBktBound(sc, value, true)
+		return
+	}
+	if idx%2 == 0 && !match {
+		bkt.updateBktBound(sc, value, false)
+	}
+}
+
+func (delta *HistDelta) mergeHistDelta(sc *stmtctx.StatementContext, rDelta *HistDelta, hg *Histogram) {
+	// Make sure that the two delta's histogram version are up to date.
+	delta.updateHistDelta(sc, hg, nil, 0)
+	rDelta.updateHistDelta(sc, hg, nil, 0)
+	for id, rBkt := range rDelta.BktDeltas {
+		bkt, ok := delta.BktDeltas[id]
+		if !ok {
+			delta.BktDeltas[id] = rBkt
+			continue
+		}
+		bkt.Count += rBkt.Count
+		if rBkt.Bound == nil {
+			continue
+		}
+		bkt.updateBktBound(sc, rBkt.Bound, id == hg.Len())
+	}
+}
+
+func (delta *TableDelta) mergeTableDelta(sc *stmtctx.StatementContext, rDelta *TableDelta, table *Table) {
+	delta.Count += rDelta.Count
+	delta.Delta += rDelta.Delta
+	if delta.IdxDeltas == nil {
+		delta.IdxDeltas = rDelta.IdxDeltas
+	} else {
+		for id, rd := range rDelta.IdxDeltas {
+			idx, ok := table.Indices[id]
```

Comment:
It seems the check in line 189 and 204 is unnecessary? Without them we also needn't table arg.

Reply:
Checking it because the corresponding index may have been deleted so that we do not need to merge them anymore. Even if without them we still need the table arg because we need the corresponding histograms.